bluemira/codes/openmc/csg_reactor.py

```diff
@@ -0,0 +1,466 @@
+
+    def make_dividing_surface(
+        self, component
+    ) -> openmc.Surface | openmc.model.ZConeOneSided:
+        """Surface that marks the end of the divertor/blanket's exterior."""
+        exterior_pts = component.exterior_vertices()[:, ::2]
+        return self.surface_from_2points(exterior_pts[0], exterior_pts[-1])
+
+    def get_blanket_and_divertor_outer_regions(
+        self, *, control_id: bool = False
+    ) -> openmc.Region:
+        """
+        Get the entire tokamak's poloidal cross-section (everything inside
+        self.geom.boundary) as an openmc.Region.
+        """
+        blanket = self.universe.blanket
+        divertor = self.universe.divertor
+        dividing_surface = self.make_dividing_surface(blanket)
+        blanket_outer = self.region_from_surface_series(
+            [*blanket.exterior_surfaces(), dividing_surface],
+            blanket.exterior_vertices(),
+            control_id=control_id,
+        )
+        divertor_outer = self.region_from_surface_series(
+            # TODO @OceanNuclear: this is wrong if the exterior surface of the divertor is not convex.  # noqa: TD003, W505, E501
+            [*divertor.exterior_surfaces(), dividing_surface],
+            divertor.exterior_vertices(),
+            control_id=control_id,
+        )
+        return blanket_outer, divertor_outer
+
+    @classmethod
+    def from_pre_cell_reactor(
+        cls,
+        pre_cell_reactor: NeutronicsReactor,
+        materials: MaterialsLibrary,
+        *,
+        control_id: bool = False,
+    ) -> CSGComponents:
+        """Make pre-cell arrays for the blanket and the divertor.
+
+        Parameters
+        ----------
+        materials:
+            library containing information about the materials
+        tokamak_dimensions:
+            A parameter
+            :class:`bluemira.radiation_transport.neutronics.params.TokamakDimensions`,
+            Specifying the dimensions of various layers in the blanket, divertor, and
+            central solenoid.
+        control_id: bool
+            Whether to set the blanket Cells and surface IDs by force or not.
+            With this set to True, it will be easier to understand where each cell came
+            from. However, it will lead to warnings and errors if a cell/surface is
+            generated to use a cell/surface ID that has already been used respectively.
+            Keep this as False if you're running openmc simulations multiple times in one
+            session.
+        """
+        # determine universe_box
+
+        z_max, z_min, r_max, r_min = pre_cell_reactor.half_bounding_box
+
+        z_min_adj = z_min - D_TOLERANCE
+        z_max_adj = z_max + D_TOLERANCE
+        r_max_adj = r_max + D_TOLERANCE
+
+        rad_shield_wall_tk = pre_cell_reactor.tokamak_dimensions.rad_shield.wall
+
+        # make the universe box, incorporates the radiation shield wall
+        self = cls(CSGComponents(*[None for _ in CSGComponents.__dataclass_fields__]))
```

Comment:
This is not a nice way to do this, it also is not correct with your typing above because your setting everything to None.

A more obvious and clean way is to just make local variables for universe, blanket etc and put them in the dataclass as the last step. There's possibly some mechanics here that make that harder but that probably should be changed.
For instance you can have post init variables that are created after the initial components are provided